maraboupy/MarabouNetworkNNet.py

```diff
@@ -263,31 +563,346 @@ def buildEquations(self):
+        else:
+            inputsNorm = inputs
+
+        # Evaluate the neural network
+        for layer in range(numLayers - 1):
+            inputsNorm = np.maximum(np.dot(weights[layer], inputsNorm) + biases[layer], 0)
+
+        if (activate_output_layer):
+            outputs = np.maximum(np.dot(weights[-1], inputsNorm) + biases[-1], 0)
+        else:
+            outputs = np.dot(weights[-1], inputsNorm) + biases[-1]
+
+        # Undo output normalization
+        if (normalize_outputs):
+            for i in range(outputSize):
+                outputs[i] = outputs[i] * self.outputRange + self.outputMean
+
+        return outputs
+
+    def evaluateNetworkMultiple(self, inputs, normalize_inputs=True, normalize_outputs=True,
+                                activate_output_layer=False):
+
+        """
+         Evaluate network using multiple sets of inputs
+
+         Args:
+             inputs (numpy array of lists of floats): Array of network inputs to be evaluated.
+
+         Returns:
+             (numpy array of numpy arrays floats): Network outputs for each set of inputs
+
+        HAS NOT BEEN TESTED
+         """
+
+        numLayers = self.numLayers
+        inputSize = self.inputSize
+        outputSize = self.outputSize
+        biases = self.biases
+        weights = self.weights
+        inputs = np.array(inputs).T
+        mins = self.inputMinimums
+        maxes = self.inputMaximums
+        means = self.inputMeans
+        ranges = self.inputRanges
+
+        # Prepare the inputs to the neural network
+        numInputs = inputs.shape[1]
+
+        if (normalize_inputs):
+            inputsNorm = np.zeros(inputSize)
+            for i in range(inputSize):
+                if inputs[i] < mins[i]:
+                    inputsNorm[i] = (mins[i] - means[i]) / ranges[i]
+                elif inputs[i] > maxes[i]:
+                    inputsNorm[i] = (maxes[i] - means[i]) / ranges[i]
+                else:
+                    inputsNorm[i] = (inputs[i] - means[i]) / ranges[i]
+        else:
+            inputsNorm = inputs
+
+        # Evaluate the neural network
+        for layer in range(numLayers - 1):
+            inputsNorm = np.maximum(np.dot(weights[layer], inputsNorm) + biases[layer].reshape((len(biases[layer]), 1)),
+                                    0)
+
+        if (activate_output_layer):
+            outputs = np.maximum(np.dot(weights[-1], inputsNorm) + biases[-1].reshape((len(biases[-1]), 1)), 0)
+        else:
+            outputs = np.dot(weights[-1], inputsNorm) + biases[-1].reshape((len(biases[-1]), 1))
+
+        # Undo output normalization
+        if (normalize_outputs):
+            for i in range(outputSize):
+                for j in range(numInputs):
+                    outputs[i, j] = outputs[i, j] * self.outputRange + self.outputMean
+
+        return outputs.T
+
+    def evaluateNetworkToLayer(self, inputs, last_layer=-1, normalize_inputs=False, normalize_outputs=False,
```

Comment:
Also seems to duplicate the previous functions - can we unify them?

Reply:
You are right. Once upon a time, it made sense to have one method with straightforward functionality that I could compare all the other ones to in order to make sure that I am not doing anything stupid while adding different functionalities, but now this is all unnecessary. I have now combined all the evaluate methods into one (with optional parameters), and removed (for now) evaluate multiple. So there should be any code duplication.

Reply:
Hi Guy, 

I think I've addressed all the points that you've raised, can you please have another quick look?